File: 26_insert_interval.js
/*
You are given an array of non-overlapping intervals intervals where 
intervals[i] = [starti, endi] represent the start and the end of the ith 
interval and intervals is sorted in ascending order by starti. 

You are also given an interval newInterval = [start, end] that represents 
the start and end of another interval.

Insert newInterval into intervals such that intervals is still sorted in 
ascending order by starti and intervals still does not have any overlapping 
intervals (merge overlapping intervals if necessary).

Return intervals after the insertion.

0 <= intervals.length <= 10^4
intervals[i].length == 2
0 <= starti <= endi <= 10^5
intervals is sorted by starti in ascending order.
newInterval.length == 2
0 <= start <= end <= 10^5
*/

// time complexity: O(n)
// space complexity: O(n)
var insert = function (intervals, newInterval) {
  let result = [];
  let [start, end] = newInterval;
  let pushed = false;
  // loop through the intervals
  for (let [currStart, currEnd] of intervals) {
    // if the current interval is less than the new interval
    if (start > currEnd) {
      result.push([currStart, currEnd]);
      // if the new interval is less than the current interval
    } else if (end < currStart) {
      result.push([start, end]);
      // push the current interval after the new interval (only once)
      if (!pushed) {
        result.push([currStart, currEnd]);
        pushed = true;
      }
      start = Infinity;
    } else {
      // merge the intervals
      start = Math.min(start, currStart);
      end = Math.max(end, currEnd);
    }
  }
  if (!pushed) result.push([start, end]);
  return result;
};

File: 26_insert_interval.py
# You are given an array of non-overlapping intervals intervals where intervals[i] = [starti, endi] represent the start and the end of the ith interval and intervals is sorted in ascending order by starti. You are also given an interval newInterval = [start, end] that represents the start and end of another interval.

# Insert newInterval into intervals such that intervals is still sorted in ascending order by starti and intervals still does not have any overlapping intervals (merge overlapping intervals if necessary).

# Return intervals after the insertion.

# Note that you don't need to modify intervals in-place. You can make a new array and return it.

# https://leetcode.com/problems/insert-interval/

# key: only append once for overlapping intervals
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        ans = []
        
        for low, hi in intervals:
            # Current interval is completely before new interval
            if hi < newInterval[0]:  
                ans.append([low, hi])
            # Current interval is completely after new interval
            elif low > newInterval[1]:  
                ans.append(newInterval)
                # Update newInterval to the remaining part
                newInterval = [low, hi] 
            # Overlapping intervals 
            else:  
                newInterval[0] = min(newInterval[0], low)
                newInterval[1] = max(newInterval[1], hi)
        
        # Add the last interval (newInterval) that hasn't been added
        ans.append(newInterval)
        
        return ans